testing/src/test/java/org/oppia/testing/FakeExceptionLoggerTest.kt

```diff
@@ -0,0 +1,194 @@
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.util.logging.ExceptionLogger
+import org.robolectric.annotation.Config
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.KClass
+import kotlin.reflect.full.cast
+
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class FakeExceptionLoggerTest {
+
+  @Inject
+  lateinit var fakeExceptionLogger: FakeExceptionLogger
+
+  @Inject
+  lateinit var exceptionLogger: ExceptionLogger
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testFakeExceptionLogger_logException_returnsException() {
+    exceptionLogger.logException(IllegalStateException("Test Exception"))
+    val exception = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exception).isInstanceOf(IllegalStateException::class.java)
+    assertThat(exception).hasMessageThat().contains("Test Exception")
+  }
+
+  @Test
+  fun testFakeExceptionLogger_logExceptionTwice_returnsLatestException() {
+    exceptionLogger.logException(IllegalStateException("First Exception"))
+    exceptionLogger.logException(NullPointerException("Second Exception"))
+    val exception = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exception).isInstanceOf(NullPointerException::class.java)
+    assertThat(exception).hasMessageThat().contains("Second Exception")
+  }
+
+  @Test
+  fun testFakeExceptionLogger_logException_clearList_logExceptionAgain_returnsLatestException() {
+    exceptionLogger.logException(IllegalStateException("First Exception"))
+    fakeExceptionLogger.clearAllExceptions()
+    exceptionLogger.logException(NullPointerException("Second Exception"))
+    val exception = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exception).isInstanceOf(NullPointerException::class.java)
+    assertThat(exception).hasMessageThat().contains("Second Exception")
+  }
+
+  @Test
+  fun testFakeExceptionLogger_logNothing_getMostRecent_returnsFailure() {
+    val exception = assertThrows(NoSuchElementException::class) {
+      fakeExceptionLogger.getMostRecentException()
+    }
+
+    assertThat(exception).isInstanceOf(NoSuchElementException::class.java)
+    assertThat(exception).hasMessageThat().contains("List is empty.")
+  }
+
+  @Test
+  fun testFakeExceptionLogger_logException_clearAllExceptions_getMostRecent_returnsFailure() {
+    exceptionLogger.logException(IllegalStateException("First Exception"))
+    fakeExceptionLogger.clearAllExceptions()
+
+    val exception = assertThrows(NoSuchElementException::class) {
+      fakeExceptionLogger.getMostRecentException()
+    }
+
+    assertThat(exception).isInstanceOf(NoSuchElementException::class.java)
+    assertThat(exception).hasMessageThat().contains("List is empty.")
+  }
+
+  @Test
+  fun testFakeExceptionLogger_clearAllExceptions_returnsEmptyList() {
+    fakeExceptionLogger.clearAllExceptions()
+    val exceptionList = fakeExceptionLogger.exceptionList
```

Comment:
We shouldn't access the implementations of the class (prefer to test the API of a class, not its implementation details). If you think it's useful for tests to be able to query whether an exception has been logged, you could introduce a method like ``hasExceptionsLogged()`` and test that. Or, you could infer that the exception list is inferred through other methods (like how ``getMostRecentException`` throws if called after ``clearAllExceptions`` is called).

The purpose of tests should be to verify that a method of the public API behaves correctly under certain conditions, not that it has a particular implementation. We should be able to switch the full implementation of a class and not change any tests if they are correctly only testing the API of the class.

Ditto elsewhere here.

Reply:
Okay, done :)